test/common/http/filter/fault_filter_test.cc

```diff
@@ -413,4 +413,46 @@ TEST_F(FaultFilterTest, FixedDelayAndAbortHeaderMatchFail) {
   EXPECT_EQ(0UL, config_->stats().aborts_injected_.value());
 }
 
+TEST_F(FaultFilterTest, TimerResetAfterStreamReset) {
+  SetUpTest(fixed_delay_only_json);
+
+  // Prep up with a 5s delay
+  EXPECT_CALL(runtime_.snapshot_, featureEnabled("fault.http.delay.fixed_delay_percent", 100))
+      .Times(1)
+      .WillOnce(Return(true));
+
+  EXPECT_CALL(runtime_.snapshot_, getInteger("fault.http.delay.fixed_duration_ms", 5000))
+      .Times(1)
+      .WillOnce(Return(5000UL));
+
+  SCOPED_TRACE("FixedDelayWithStreamReset");
+  timer_ = new Event::MockTimer(&filter_callbacks_.dispatcher_);
+  EXPECT_CALL(*timer_, enableTimer(std::chrono::milliseconds(5000UL)));
+
+  EXPECT_CALL(filter_callbacks_.request_info_,
+              setResponseFlag(Http::AccessLog::ResponseFlag::DelayInjected)).Times(1);
+
+  EXPECT_EQ(0UL, config_->stats().delays_injected_.value());
+
+  EXPECT_EQ(FilterHeadersStatus::StopIteration, filter_->decodeHeaders(request_headers_, false));
+
+  // delay timer should have been fired by now. If caller resets the stream while we are waiting
+  // on the delay timer, check if timers are cancelled
+  EXPECT_CALL(*timer_, disableTimer());
+
+  // The timer callback should never be called.
+  EXPECT_CALL(runtime_.snapshot_, featureEnabled("fault.http.abort.abort_percent", _)).Times(0);
+  EXPECT_CALL(runtime_.snapshot_, getInteger("fault.http.abort.http_status", _)).Times(0);
+  EXPECT_CALL(filter_callbacks_, encodeHeaders_(_, _)).Times(0);
+  EXPECT_CALL(filter_callbacks_.request_info_,
+              setResponseFlag(Http::AccessLog::ResponseFlag::FaultInjected)).Times(0);
+  EXPECT_CALL(filter_callbacks_, continueDecoding()).Times(0);
+  EXPECT_EQ(0UL, config_->stats().aborts_injected_.value());
+
+  EXPECT_EQ(FilterDataStatus::Continue, filter_->decodeData(data_, true));
```

Comment:
Though it works from a "test this thing" perspective, this is not a totally realistic test case. What you are doing here is calling decodeData with end_stream = true. This means that the request is over, as such it's impossible for decodeTrailers() to be called, so I would delete line 453. FYI, it is possible for resetStream callback to get called even when request is complete, since downstream can reset waiting for a response.